### PR TITLE
STITCH-1039 JS SDK: Improve semantics of multiple log ins

### PR DESCRIPTION
**Pull request by jsflax**

### CODE REVIEW

File: src/auth/index.js [hunk 1]
```diff
@@ -67,7 +67,8 @@ export default class Auth {
   }
 
   refreshToken() {
-    return this.client.doSessionPost().then(json => this.set(json));
+    return Promise.all([this.client.doSessionPost(), this.getLoggedInProviderType()])
```
Comment (edaniels) [inline src/auth/index.js]: [q] Why does refreshing a token set the provider type?

File: src/auth/index.js [hunk 2]
```diff
@@ -123,7 +124,7 @@ export default class Auth {
       }
 
       // If we get here, the state is valid - set auth appropriately.
-      return this.set(redirectState.ua);
+      return this.getLoggedInProviderType().then(providerType => this.set(redirectState.ua, providerType));
```
Comment (edaniels) [inline src/auth/index.js]: [major] How does the ordering work here? `getLoggedInProviderType` wouldn't be able to return anything useful when it's the first time we're here and as a result, `set` will fail due to the restriction it has in place about empty types.

File: src/auth/index.js [hunk 3]
```diff
@@ -160,14 +161,18 @@ export default class Auth {
 
     document.cookie = `${authCommon.USER_AUTH_COOKIE_NAME}=; Path=/; Expires=Thu, 01 Jan 1970 00:00:00 GMT;`;
     const userAuth = this.unmarshallUserAuth(uaCookie);
-    return this.set(userAuth).then(() =>
+    return this.getLoggedInProviderType().then(providerType => this.set(userAuth, providerType)).then(() =>
```
Comment (edaniels) [inline src/auth/index.js]: [major] How does the ordering work here? `getLoggedInProviderType` wouldn't be able to return anything useful when it's the first time we're here and as a result, `set` will fail due to the restriction it has in place about empty types.

File: src/auth/index.js [hunk 4]
```diff
@@ -208,19 +213,21 @@ export default class Auth {
     return this.storage.get(authCommon.REFRESH_TOKEN_KEY);
   }
 
-  set(json) {
-    if (!json) {
+  set(json, authType) {
+    if (!json || !authType) {
```
Comment (edaniels) [inline src/auth/index.js]: It looks like there are cases where `authType` could definitely be empty which would cause us to not complete a log in. Is it necessary that we care what `authType` is?

File: src/auth/providers.js [hunk 5]
```diff
@@ -7,6 +7,8 @@ import { uriEncodeObject } from '../util';
  * @namespace
  */
 function anonProvider(auth) {
+  const providerType = 'anon';
```
Comment (edaniels) [inline src/auth/providers.js]: [nit] We should probably make this exactly the same name/type that we use in the route for every provider. i.e. `anon` -> `anon-user`.

File: src/client.js [hunk 6]
```diff
@@ -136,14 +136,29 @@ export default class StitchClient {
...
+    ).then(([accessToken, loggedInProviderType]) => {
       if (accessToken) {
-        return this.auth.authedId();
+        // TODO: `anon` should be enumerated once providers are maintain
+        // an interface
+        if (providerType === 'anon' && loggedInProviderType === 'anon') {
+          return false; // is authenticated, skip log in
+        }
+
+        return this.logout(); // will not be authenticated, continue log in
```
Comment (edaniels) [inline src/client.js]: [minor] If logout fails, then we will never authenticate. From ticket:
`the SDK should attempt to log out (failure should be swallowed), and logged in with the new provider`. We should have a test case for this.

File: src/client.js [hunk 7]
```diff
@@ -136,14 +136,29 @@ export default class StitchClient {
...
-      return this.auth.provider(providerType).authenticate(options)
-        .then(() => this.auth.authedId());
-    });
+      return true; // is not authenticated, continue log in
+    }).then((isNotAuthed) => {
+      if (isNotAuthed) {
+        return this.auth.provider(providerType).authenticate(options);
+      }
+      return;
+    }).then(() => this.auth.authedId());
```
Comment (edaniels) [inline src/client.js]: [q] Can just pass `this.auth.authedId` instead of having an arrow function, right?

File: test/auth.test.js [hunk 8]
```diff
@@ -26,18 +29,24 @@ function mockApiResponse(options = {}) {
...
+  });
 
   it('should return a promise for anonymous login with existing auth data', async() => {
     window.fetch.resolves(mockApiResponse());
     expect.assertions(1);
 
     let client = new StitchClient();
     await client.auth.storage.set(common.USER_AUTH_KEY, mockAuthData());
-
+    await client.auth.storage.set(common.USER_LOGGED_IN_PT_KEY, 'anon');
```
Comment (edaniels) [inline test/auth.test.js]: [nit] This should probably use a constant

File: test/auth.test.js [hunk 9]
```diff
@@ -58,15 +67,48 @@ describe('Auth', () => {
...
+  beforeEach(async() => {
+    const { apiKey, groupId, serverUrl } = extractTestFixtureDataPoints(test);
+    th = await buildClientTestHarness(apiKey, groupId, serverUrl);
+    await th.configureAnon();
+    client = th.stitchClient;
+  });
+
+  afterEach(async() => await th.cleanup());
+
+  it('should track currently logged in provider type', async() => {
```
Comment (edaniels) [inline test/auth.test.js]: [minor] Per the ticket, we need a `isAuthenticated` like method to maintain parity with other SDKs. We should have a test that checks a flow of:
1. Verify `isAuthenticated` resolves to false
2. Log in
3. Verify `isAuthenticated` resolves to true

Comment (edaniels) [inline test/auth.test.js]: [minor] We should assert a logout did indeed happen besides just re logging in.

File: src/client.js [hunk 10]
```diff
@@ -136,14 +136,29 @@ export default class StitchClient {
...
-    return this.auth.getAccessToken().then(accessToken => {
+    return Promise.all(
+      [
+        this.auth.getAccessToken(),
+        this.auth.getLoggedInProviderType()
+      ]
+    ).then(([accessToken, loggedInProviderType]) => {
       if (accessToken) {
-        return this.auth.authedId();
+        // TODO: `anon` should be enumerated once providers are maintain
```
Comment (edaniels) [inline src/client.js]: [nit] Instead of enumeration, we could just have a constant for now

File: src/auth/providers.js [hunk 11]
```diff
@@ -292,7 +298,7 @@ function googleProvider(auth) {
         })
           .then(common.checkStatus)
           .then(response => response.json())
-          .then(json => auth.set(json));
+          .then(json => auth.set(json, providerType));
       }
 
       const redirectUrl = (data && data.redirectUrl) ? data.redirectUrl : undefined;
```
Comment (edaniels) [inline src/auth/providers.js]: [minor] Redirects are not being handled for provider types being prospectively set in advance. You may be able to set a temporary field related to the request like STATE_KEY

Comment (edaniels) [reply src/auth/providers.js]: [major] Speaking of `STATE_KEY`, I'm pretty convinced `window.location.replace(getOAuthLoginURL(auth, 'google', redirectUrl));` and the facebook one are broken with the async changes to storage. getOAuthLoginURL needs to be thenned before using window location replace.

File: src/auth/index.js (hunk 3 above)
Comment (edaniels) [reply src/auth/index.js]: This one isn't like redirect and in this special case, we can assume the providerType is `mongodb-cloud`.

File: src/client.js [hunk 12]
```diff
@@ -136,14 +136,29 @@ export default class StitchClient {
    */
   authenticate(providerType, options = {}) {
     // reuse existing auth if present
-    return this.auth.getAccessToken().then(accessToken => {
+    return Promise.all(
+      [
+        this.auth.getAccessToken(),
```
Comment (edaniels) [inline src/client.js]: [nit] It seems like it would be better to use just `getLoggedInProviderType` here right? You're only acting upon info from `providerType`

**Comment by edaniels** [comment]
Evergreen tests are failing. Might be legit

File: src/auth/index.js [hunk 13]
```diff
@@ -264,6 +271,10 @@ export default class Auth {
     });
   }
 
+  getLoggedInProviderType() {
+    return this.storage.get(authCommon.USER_LOGGED_IN_PT_KEY);
```
Comment (edaniels) [inline src/auth/index.js]: [minor] Given that this directly proxies a storage async method, we should probably be better about the resolved output regardless of failure conditions. Under normal circumstances, this should return the type as a string, otherwise, an empty string could suffice. Another option is to throw an exception if `isAuthenticated` resolves to false.

File: src/client.js (hunk 12 above)
Comment (edaniels) [reply src/client.js]: Now that I think about it, it seems like we actually want to check `isAuthenticated` and then `getLoggedInProviderType`. The thing to keep in mind is that we only want to call `getLoggedInProviderType` if it's going to not reject or anything. Otherwise we drop out of the next then and never authenticate.

File: src/auth/index.js (hunk 4 above)
Comment (jsflax) [reply src/auth/index.js]: What is a case where `authType` could be empty? We don't care what `authType` is, but we care if it's `null`. It should be impossible to log in to stitch without a provider type.

File: src/auth/index.js (hunk 13 above)
Comment (jsflax) [reply src/auth/index.js]: An exception could make sense, otherwise I think an `undefined` or `null` is an accurate representation of there not being a provider type.

File: src/client.js (hunk 12 above)
Comment (jsflax) [reply src/client.js]: `isAuthenticated` doesn't exist exactly– it would be a mix of checking these fields anyway.

File: src/auth/index.js (hunk 4 above)
Comment (edaniels) [reply src/auth/index.js]: The handling of cookies and redirects

File: src/client.js (hunk 10 above)
Comment (jsflax) [reply src/client.js]: We could, but they'd exist in an exported global state. Otherwise we'd have to heavily refactor `providers`

File: src/client.js [hunk 14]
```diff
@@ -160,7 +175,8 @@ export default class StitchClient {
         useRefreshToken: true,
         rootURL: this.rootURLsByAPIVersion[v2][API_TYPE_CLIENT]
       }
-    ).then(() => this.auth.clear()); // eslint-disable-line space-before-function-paren
+    ).then(() => this.auth.clear())
+      .then(() => true);
```
Comment (mpobrien) [inline src/client.js]: q: why is this needed?

File: src/client.js (hunk 12 above)
Comment (edaniels) [reply src/client.js]: I think it should exist. Users need to have a decent way of checking if they are authed before doing anything. authedId seems roundabout

File: test/auth.test.js [hunk 15]
```diff
@@ -26,18 +29,24 @@ function mockApiResponse(options = {}) {
...
 describe('Auth', () => {
-  beforeEach(() => { test.fetch = sinon.stub(window, 'fetch'); });
-  afterEach(() => test.fetch.restore());
+  let test = {};
+
+  beforeEach(() => {
+    test.fetch = sinon.stub(window, 'fetch');
+  });
+
+  afterEach(async() => {
```
Comment (mpobrien) [inline test/auth.test.js]: nit: this doesn't need to be `async` if we're not using `await` in the body

File: src/client.js (hunk 12 above)
Comment (edaniels) [reply src/client.js]: Also, consider someone with existing local storage or a new SDK.  Won't `authenticate` ultimately reject because they have nothing set.

File: src/client.js (hunk 7 above)
Comment (jsflax) [reply src/client.js]: Probably. Will have to try.

File: test/auth.test.js (hunk 9 above)
Comment (jsflax) [reply test/auth.test.js]: I assert this in a later test.

Comment (jsflax) [reply test/auth.test.js]: `isAuthenticated` is going to be a wrapper over checking `authedId`. It seems weird to have multiple sources of truth for this, but I can add the method.

File: src/client.js (hunk 10 above)
Comment (edaniels) [reply src/client.js]: Why would it be global if you're just importing the definitions?

File: src/auth/index.js (hunk 13 above)
Comment (edaniels) [reply src/auth/index.js]: Yeah so if we make it an exception, we will have to be very careful using it with Promise.all by not using it all and instead chaining it with `isAuthenticated` and then `getLoggedInProviderType`. `undefined` or `null` is okay too.

File: src/auth/providers.js (hunk 5 above)
Comment (edaniels) [reply src/auth/providers.js]: I take this back. We should really haver a constant here that is used in createProviders as well.

File: test/auth.test.js (hunk 9 above)
Comment (edaniels) [reply test/auth.test.js]: I don't see that the other test actually asserts that we sent a logout to the server instead of re logging in.

Comment (edaniels) [reply test/auth.test.js]: I'd argue that `authedId` should throw if you're not authed.

File: src/auth/providers.js [hunk 16]
```diff
@@ -267,6 +272,7 @@ function getOAuthLoginURL(auth, providerName, redirectUrl) {
 /** @namespace */
 function googleProvider(auth) {
   const loginRoute = auth.isAppClient() ? 'providers/oauth2-google/login' : 'providers/oauth2-google/login';
+  const providerType = 'google';
```
Comment (mpobrien) [inline src/auth/providers.js]: we should declare these special strings in constants up top. we use em all over (including in tests)

File: test/auth.test.js [hunk 17]
```diff
@@ -58,15 +67,48 @@ describe('Auth', () => {
...
+    await client.login(email, password);
+    expect(await client.auth.storage.get(common.USER_LOGGED_IN_PT_KEY)).toEqual('userpass');
+  });
+
+  it('should return existing login for relogging "anon"', async() => {
+    const userId = await client.login();
+    expect(userId).toEqual(await client.login());
+  });
+
+  it('should login with a new provider if switching providers', async() => {
```
Comment (mpobrien) [inline test/auth.test.js]: it looks like we sorta already do this in one of the previous tests at L94-95. Maybe we should just nest this `it` block into there?

File: src/auth/index.js (hunk 1 above)
Comment (jsflax) [reply src/auth/index.js]: These functions were locked together and would've required a larger refactor.

Comment (edaniels) [reply src/auth/index.js]: Gotcha 👍

File: src/client.js (hunk 6 above)
Comment (jsflax) [reply src/client.js]: I'll add a test– the code above should work to spec.

File: src/client.js (hunk 14 above)
Comment (jsflax) [reply src/client.js]: As a successful return value. `clear()` doesn't produce anything useful– it would be nice to have a primitive value to check on the success of the logout.

File: test/auth.test.js (hunk 9 above)
Comment (jsflax) [reply test/auth.test.js]: Didn't assert, but it's checked implicitly in `should re-login with the same provider if not anon`. I'll add the assertion.

File: test/auth.test.js (hunk 17 above)
Comment (jsflax) [reply test/auth.test.js]: Sure, I don't have a strong opinion here. This test was just more explicit.

File: test/auth.test.js (hunk 15 above)
Comment (jsflax) [reply test/auth.test.js]: Good catch, leftover artifact from an earlier version of this.

File: src/client.js (hunk 10 above)
Comment (jsflax) [reply src/client.js]: Right, I guess just because it isn't bound to any class or function doesn't mean it's global if it's exported properly and is stateless. I'm fine with making these constants, but it will be much nicer when each provider is implementing an interface that owns this const.

File: src/client.js (hunk 14 above)
Comment (mpobrien) [reply src/client.js]: but the logout is implicitly successful if the promise that it returns doesn't reject, which the caller can check for with `.then`/`.catch`. It doesn't look like we use the `true` value anyway, so i don't see the point

File: src/client.js (hunk 7 above)
Comment (jsflax) [reply src/client.js]: Apparently not, unless I'm doing it wrong.

File: src/client.js (hunk 14 above)
Comment (jsflax) [reply src/client.js]: We use it in the new authenticate method.

File: src/client.js (hunk 7 above)
Comment (edaniels) [reply src/client.js]: nah makes sense, the this here is the client

File: src/client.js [hunk 18]
```diff
@@ -136,14 +137,27 @@ export default class StitchClient {
...
+
+        return this.logout(); // will not be authenticated, continue log in
       }
 
-      return this.auth.provider(providerType).authenticate(options)
-        .then(() => this.auth.authedId());
-    });
+      return true; // is not authenticated, continue log in
+    }).then((isNotAuthed) => {
+      if (isNotAuthed) {
```
Comment (edaniels) [inline src/client.js]: [nit] Feels like the name of `isNotAuthed` is more apt as `shouldAuth`.

File: src/client.js [hunk 19]
```diff
@@ -160,7 +174,8 @@ export default class StitchClient {
         useRefreshToken: true,
         rootURL: this.rootURLsByAPIVersion[v2][API_TYPE_CLIENT]
       }
-    ).then(() => this.auth.clear()); // eslint-disable-line space-before-function-paren
+    ).then(() => this.auth.clear())
```
Comment (edaniels) [inline src/client.js]: [minor] I think we need a `catch` or a second arg to `then` such that in both cases we succeed and always return with `this.auth.clear`. We can log the failure.

File: src/client.js [hunk 20]
```diff
@@ -136,14 +137,27 @@ export default class StitchClient {
...
+        this.auth.getLoggedInProviderType()
+      ]
+    ).then(([accessToken, loggedInProviderType]) => {
       if (accessToken) {
-        return this.auth.authedId();
+        if (providerType === PROVIDER_TYPE_ANON && loggedInProviderType === PROVIDER_TYPE_ANON) {
+          return false; // is authenticated, skip log in
+        }
+
+        return this.logout(); // will not be authenticated, continue log in
```
Comment (edaniels) [inline src/client.js]: [minor] I think logout we really don't care if it fails or not, just that we try and wait to do it. Given my comment on the logout method holds, we can just do `this.logout().then(() => true)`

File: src/client.js [hunk 21]
```diff
@@ -183,6 +198,14 @@ export default class StitchClient {
     )
       .then(response => response.json());
   }
+
+  /**
+  * @return {Promise} whether or not the current client is authenticated
+  */
+  isAuthenticated() {
+    return this.auth.authedId().then((id) => id != null && id != undefined);
```
Comment (edaniels) [inline src/client.js]: [minor] This promise can reject. If it's suppose to return a bool, we should probably satisfy the `reject` case on `then`.
[test] Test for when it rejects

File: test/auth.test.js [hunk 22]
```diff
@@ -1,6 +1,11 @@
 import sinon from 'sinon';
 import StitchClient from '../src/client';
 import * as common from '../src/auth/common';
+import { PROVIDER_TYPE_ANON, PROVIDER_TYPE_USERPASS } from '../src/auth/providers';
+
+const StitchMongoFixture = require('./fixtures/stitch_mongo_fixture');
```
Comment (edaniels) [inline test/auth.test.js]: [nit] `import StitchMongoFixture from './fixtures/stitch_mongo_fixture'`
[style] collapse lines

File: test/auth.test.js [hunk 23]
```diff
@@ -58,15 +67,57 @@ describe('Auth', () => {
...
+describe('Auth login semantics', () => {
+  const email = 'test_user@domain.com';
+  const password = 'password';
+  let test = new StitchMongoFixture();
+  let th;
+  let client;
 
-    let client = new StitchClient();
-    await client.auth.storage.set(common.USER_AUTH_KEY, mockAuthData());
+  beforeAll(async() => await test.setup());
```
Comment (edaniels) [inline test/auth.test.js]: [q] If `test.setup` is a promise, can't you just do `beforeAll(test.setup())`?

File: test/auth.test.js [hunk 24]
```diff
@@ -58,15 +67,57 @@ describe('Auth', () => {
...
+  afterAll(async() => await test.teardown());
```
Comment (edaniels) [inline test/auth.test.js]: Same question to all uses of before* and after* where it's a single line return

File: src/auth/index.js [hunk 25]
```diff
@@ -208,19 +213,28 @@ export default class Auth {
...
   }
 
-  set(json) {
+  setAuthType(authType) {
+    return new Promise((resolve) =>
+      resolve(this.storage.set(authCommon.USER_LOGGED_IN_PT_KEY, authType))
+    );
+  }
+
+  set(json, authType) {
```
Comment (mburdette) [inline src/auth/index.js]: you could use a default value `authType = ''` and then not do the `!authType` check in the promise that @edaniels commented on

File: src/auth/index.js [hunk 26]
```diff
@@ -264,6 +278,10 @@ export default class Auth {
     });
   }
 
+  getLoggedInProviderType() {
+    return this.storage.get(authCommon.USER_LOGGED_IN_PT_KEY).then(type => !type ? '' : type);
```
[the remark this answers is not in the document]
Comment (mburdette) [reply src/auth/index.js]: [nit] even easier would be `type || ''`

File: src/auth/index.js [hunk 27]
```diff
@@ -208,19 +213,28 @@ export default class Auth {
...
+
+  set(json, authType) {
     if (!json) {
       return;
     }
 
     let newUserAuth = {};
     return new Promise((resolve) => {
+      if (!authType) { resolve(); }
+      else { resolve(this.storage.set(authCommon.USER_LOGGED_IN_PT_KEY, authType)); }
```
Comment (mpobrien) [inline src/auth/index.js]: I think you could just use `return` statements here and not need to involve `resolve` altogether.

File: src/client.js [hunk 28]
```diff
@@ -136,14 +137,27 @@ export default class StitchClient {
...
 
-      return this.auth.provider(providerType).authenticate(options)
-        .then(() => this.auth.authedId());
-    });
+      return true; // is not authenticated, continue log in
+    }).then((isNotAuthed) => {
+      if (isNotAuthed) {
+        return this.auth.provider(providerType).authenticate(options);
+      }
+      return;
```
Comment (mpobrien) [inline src/client.js]: superfluous `return` here

File: src/auth/index.js (hunk 26 above)
[the remark this answers is not in the document]
Comment (edaniels) [reply src/auth/index.js]: ^ If you're a masochist

File: src/client.js [hunk 29]
```diff
@@ -136,14 +137,27 @@ export default class StitchClient {
    */
   authenticate(providerType, options = {}) {
     // reuse existing auth if present
-    return this.auth.getAccessToken().then(accessToken => {
+    return Promise.all(
+      [
+        this.auth.getAccessToken(),
```
[the remark this answers is not in the document]
Comment (jsflax) [reply src/client.js]: I didn't realize it rejected if nil. I'll move it around.

File: src/client.js (hunk 20 above)
Comment (jsflax) [reply src/client.js]: good point. will do

File: src/client.js (hunk 18 above)
Comment (jsflax) [reply src/client.js]: Fair point.

File: src/client.js (hunk 19 above)
Comment (jsflax) [reply src/client.js]: Good call

File: test/auth.test.js (hunk 22 above)
Comment (jsflax) [reply test/auth.test.js]: Yep, will do.

File: src/auth/index.js [hunk 30]
```diff
@@ -208,19 +213,28 @@ export default class Auth {
...
+  }
+
+  set(json, authType) {
     if (!json) {
       return;
     }
 
     let newUserAuth = {};
     return new Promise((resolve) => {
+      if (!authType) { resolve(); }
```
[the remark this answers is not in the document]
Comment (jsflax) [reply src/auth/index.js]: Will address with @mburdette 's comment

File: test/testutil.js [hunk 31]
```diff
@@ -51,6 +51,12 @@ class TestHarness {
     });
   }
 
+  async configureAnon() {
```
[the remark this answers is not in the document]
Comment (jsflax) [reply test/testutil.js]: will remove

File: test/auth.test.js (hunk 24 above)
Comment (jsflax) [reply test/auth.test.js]: fixing

File: src/auth/index.js (hunk 25 above)
Comment (jsflax) [reply src/auth/index.js]: done

File: src/auth/index.js (hunk 27 above)
Comment (jsflax) [reply src/auth/index.js]: done

File: src/client.js (hunk 28 above)
Comment (jsflax) [reply src/client.js]: done

File: test/auth.test.js (hunk 23 above)
Comment (jsflax) [reply test/auth.test.js]: Apparently, that isn't allowed.

File: test/auth.test.js (hunk 24 above)
Comment (jsflax) [reply test/auth.test.js]: Apparently, that isn't allowed.

Comment (edaniels) [reply test/auth.test.js]: why not?

Comment (jsflax) [reply test/auth.test.js]:
```
encountered a declaration exception
TypeError: fn.call is not a function
      
      at Object.asyncFn (node_modules/jest-jasmine2/build/jasmine-async.js:38:30)
          at Promise (<anonymous>)
      at Promise.resolve.then.el (node_modules/p-map/index.js:46:16)
          at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:188:7)
```

It's expecting to be wrapped in a block by the looks of it

**Comment by jsflax** [comment]
@edaniels @mpobrien @mburdette PTAL

File: test/auth.test.js (hunk 24 above)
Comment (edaniels) [reply test/auth.test.js]: what code are you passing it?

File: src/auth/index.js [hunk 32]
```diff
@@ -214,27 +214,25 @@ export default class Auth {
...
 
     let newUserAuth = {};
-    return new Promise((resolve) => {
-      if (!authType) { resolve(); }
-      else { resolve(this.storage.set(authCommon.USER_LOGGED_IN_PT_KEY, authType)); }
-    }).then(() => {
+
+    return (authType ? 
+      this.storage.set(authCommon.USER_LOGGED_IN_PT_KEY, authType) : 
+      new Promise()
```
Comment (mpobrien) [inline src/auth/index.js]: I think this needs to be `Promise.resolve()`, rather than `new Promise()`.

File: test/auth.test.js (hunk 24 above)
Comment (jsflax) [reply test/auth.test.js]: `beforeAll(test.setup())` / the equivalent for each block

File: src/client.js (hunk 28 above)
Comment (mpobrien) [reply src/client.js]: i think it's still there!

File: src/auth/index.js (hunk 32 above)
Comment (jsflax) [reply src/auth/index.js]: `new Promise()` has been tested and works. But if you'd prefer `Promise.resolve()`, I can alter it.

File: test/auth.test.js (hunk 24 above)
Comment (edaniels) [reply test/auth.test.js]: darn. oh well

File: src/auth/index.js (hunk 32 above)
Comment (mpobrien) [reply src/auth/index.js]: Are you sure? I think you're required to pass a function to the `Promise` constructor otherwise there's no way for it to hook into the following `.then`/`.catch`.
e.g. if i do that I get this:
```
> (new Promise()).then(()=>{console.log("hi")})
TypeError: Promise resolver undefined is not a function
    at Promise (<anonymous>)
```

if i use `Promise.resolve` though i get:
```> (Promise.resolve()).then(()=>{console.log("hi")})
Promise {
  <pending>,
  domain:
   Domain {
     domain: null,
     _events: { error: [Function: debugDomainError] },
     _eventsCount: 1,
     _maxListeners: undefined,
     members: [] } }
> hi
```

Comment (jsflax) [reply src/auth/index.js]: Fixed

File: src/auth/index.js [hunk 33]
```diff
@@ -67,7 +67,8 @@ export default class Auth {
   }
 
   refreshToken() {
-    return this.client.doSessionPost().then(json => this.set(json));
+    return Promise.all([this.client.doSessionPost(), this.getLoggedInProviderType()])
+      .then(([json, providerType]) => this.set(json, providerType));
```
Comment (edaniels) [inline src/auth/index.js]: Remove `Promise.all`
